File: frontend/src/api/rooms.ts
import { apiClient } from "./apiUtils";

export type RoomChannelType = "slack" | "webhook";

export interface RoomSummary {
  id: number;
  name: string;
  description?: string;
  member_count: number;
  channel_types: RoomChannelType[];
}

export interface Room {
  id: number;
  team_id: number;
  name: string;
  description?: string;
  created_at: string;
  updated_at: string;
}

export interface RoomMemberDetail {
  room_id: number;
  user_id: number;
  name?: string;
  email?: string;
  role: string;
  added_at: string;
}

export interface RoomChannel {
  id: number;
  room_id: number;
  type: RoomChannelType;
  name?: string;
  config: Record<string, any>;
  enabled: boolean;
  created_at: string;
  updated_at: string;
}

export interface CreateRoomRequest {
  name: string;
  description?: string;
}

export interface UpdateRoomRequest {
  name?: string;
  description?: string;
}

export interface AddRoomMemberRequest {
  user_id: number;
  role: string;
}

export interface CreateRoomChannelRequest {
  type: RoomChannelType;
  name?: string;
  config: Record<string, any>;
  enabled?: boolean;
}

export interface UpdateRoomChannelRequest {
  name?: string;
  config?: Record<string, any>;
  enabled?: boolean;
}

export const roomsApi = {
  listRooms: (teamId: number) =>
    apiClient.get<RoomSummary[]>(`/teams/${teamId}/rooms`),
  createRoom: (teamId: number, payload: CreateRoomRequest) =>
    apiClient.post<Room>(`/teams/${teamId}/rooms`, payload),
  updateRoom: (teamId: number, roomId: number, payload: UpdateRoomRequest) =>
    apiClient.put<Room>(`/teams/${teamId}/rooms/${roomId}`, payload),
  deleteRoom: (teamId: number, roomId: number) =>
    apiClient.delete<{ message: string }>(`/teams/${teamId}/rooms/${roomId}`),

  listMembers: (teamId: number, roomId: number) =>
    apiClient.get<RoomMemberDetail[]>(`/teams/${teamId}/rooms/${roomId}/members`),
  addMember: (teamId: number, roomId: number, payload: AddRoomMemberRequest) =>
    apiClient.post<{ message: string }>(`/teams/${teamId}/rooms/${roomId}/members`, payload),
  removeMember: (teamId: number, roomId: number, userId: number) =>
    apiClient.delete<{ message: string }>(`/teams/${teamId}/rooms/${roomId}/members/${userId}`),

  listChannels: (teamId: number, roomId: number) =>
    apiClient.get<RoomChannel[]>(`/teams/${teamId}/rooms/${roomId}/channels`),
  createChannel: (teamId: number, roomId: number, payload: CreateRoomChannelRequest) =>
    apiClient.post<RoomChannel>(`/teams/${teamId}/rooms/${roomId}/channels`, payload),
  updateChannel: (teamId: number, roomId: number, channelId: number, payload: UpdateRoomChannelRequest) =>
    apiClient.put<RoomChannel>(`/teams/${teamId}/rooms/${roomId}/channels/${channelId}`, payload),
  deleteChannel: (teamId: number, roomId: number, channelId: number) =>
    apiClient.delete<{ message: string }>(`/teams/${teamId}/rooms/${roomId}/channels/${channelId}`)
};

File: pkg/models/rooms.go
package models

import "time"

type RoomID int64

// Room represents a reusable notification group scoped to a team.
type Room struct {
	ID          RoomID   `json:"id"`
	TeamID      TeamID   `json:"team_id"`
	Name        string   `json:"name"`
	Description string   `json:"description,omitempty"`
	CreatedAt   time.Time `json:"created_at"`
	UpdatedAt   time.Time `json:"updated_at"`
}

// RoomMember connects a user to a room.
type RoomMember struct {
	RoomID  RoomID   `json:"room_id"`
	UserID  UserID   `json:"user_id"`
	Role    string   `json:"role"`
	AddedAt time.Time `json:"added_at"`
}

type RoomMemberDetail struct {
	RoomID  RoomID   `json:"room_id"`
	UserID  UserID   `json:"user_id"`
	Name    string   `json:"name"`
	Email   string   `json:"email"`
	Role    string   `json:"role"`
	AddedAt time.Time `json:"added_at"`
}

type RoomChannelType string

const (
	RoomChannelSlack   RoomChannelType = "slack"
	RoomChannelWebhook RoomChannelType = "webhook"
)

// RoomChannel defines a non-email delivery target for a room.
type RoomChannel struct {
	ID        int64            `json:"id"`
	RoomID    RoomID           `json:"room_id"`
	Type      RoomChannelType  `json:"type"`
	Name      string           `json:"name,omitempty"`
	Config    map[string]any   `json:"config"`
	Enabled   bool             `json:"enabled"`
	CreatedAt time.Time        `json:"created_at"`
	UpdatedAt time.Time        `json:"updated_at"`
}

// RoomSummary is a lightweight representation suitable for alerts.
type RoomSummary struct {
	ID           RoomID            `json:"id"`
	Name         string            `json:"name"`
	Description  string            `json:"description,omitempty"`
	MemberCount  int               `json:"member_count"`
	ChannelTypes []RoomChannelType `json:"channel_types"`
}

// AlertHistoryRoomSnapshot captures the room state at trigger time.
type AlertHistoryRoomSnapshot struct {
	RoomID       RoomID            `json:"room_id"`
	Name         string            `json:"name"`
	ChannelTypes []RoomChannelType `json:"channel_types"`
	MemberEmails []string          `json:"member_emails,omitempty"`
}

type CreateRoomRequest struct {
	Name        string `json:"name"`
	Description string `json:"description"`
}

type UpdateRoomRequest struct {
	Name        *string `json:"name"`
	Description *string `json:"description"`
}

type AddRoomMemberRequest struct {
	UserID UserID `json:"user_id"`
	Role   string `json:"role"`
}

type CreateRoomChannelRequest struct {
	Type    RoomChannelType `json:"type"`
	Name    string          `json:"name"`
	Config  map[string]any  `json:"config"`
	Enabled bool            `json:"enabled"`
}

type UpdateRoomChannelRequest struct {
	Name    *string          `json:"name"`
	Config  *map[string]any  `json:"config"`
	Enabled *bool            `json:"enabled"`
}
